src/module_4/push_exploration_non_linear.md

```diff
@@ -0,0 +1,2005 @@
+
+## Import libraries
+
+
+```python
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+import seaborn as sns
+import logging
+from datetime import timedelta, datetime
+
+from sklearn.preprocessing import StandardScaler
+from sklearn.linear_model import LogisticRegression
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import roc_curve, auc, precision_recall_curve
+from sklearn.metrics import confusion_matrix, classification_report
+from sklearn.pipeline import Pipeline
+from sklearn.model_selection import GridSearchCV
+import joblib
+import pickle
+import os
+import json
+from typing import List, Dict, Optional
+from catboost import Pool
+from catboost import CatBoostClassifier
+import traceback
+
+```
+
+# Functions used in previous MVP
+
+
+```python
+def assess_NA(data: pd.DataFrame):
+    """
+    Returns a pd.DataFrame denoting the total number of NA
+    values and the percentage of NA values in each column.
+    """
+    # pd.Datadenoting features and the sum of their null values
+    nulls = data.isnull().sum().reset_index().rename(columns={0: "count"})
+    nulls["percent"] = nulls["count"] * 100 / len(data)
+
+    return nulls
+
+
+def read_data(file_path: str) -> pd.DataFrame:
+    """
+    Read and validate data from a CSV file.
+    Parameters: file_path (str)
+    Returns: pd.DataFrame
+    """
+    try:
+        data = pd.read_csv(file_path, index_col=False)
+
+        if data is not None and not data.empty:
+            logging.info("Data loaded successfully.")
+
+            # Info about its shape and nulls
+            rows, cols = data.shape
+            logging.info(f"Data shape: {rows} rows x {cols} columns")
+
+            null_assessment = assess_NA(data)
+            logging.info(
+                "Assessment of NA values:\n" + null_assessment.to_string(index=False)
+            )
+
+            # Show data sample
+            print(data.head(5))
+
+            return data
+        else:
+            logging.error("Error: The loaded data is empty or None")
+            return
+
+    except FileNotFoundError as e:
+        logging.error(f"Error: The CSV file is empty. Details: {e}")
+    except pd.errors.EmptyDataError as e:
+        logging.error(
+            f"Error: An unexpected error occurred while loading the data. Details {e}"
+        )
+
+
+def preprocess_data(
+    data: pd.DataFrame, remove_if_all_na: bool = False, num_items: int = 5
+) -> pd.DataFrame:
+    """
+    Preprocess data by removing rows with
+    filter orders with at least 5 items
+
+    Parameters:
+    - data (pd.DataFrame)
+    - remove_all_na_rows (bool): If True, remove rows where at least
+    1 value is missing. If False, remove rows where all values are missing.
+
+    Returns:
+    - pd.DataFrame. The preprocessed dataset
+    """
+    try:
+        initial_length = len(data)
+        if remove_if_all_na:  # remove if everything is NA
```

Comment:
Ah okey, here you are doing the opposite that you said in the comments. However, as said before, I'd always delete rows with ALL nas and have a variable `drop_rows_with_some_na` which drops columns with at least 1 NA or not